### ISSUE TITLE
Debug models from SecureNN paper

### ISSUE THREAD

**Issue by yanndupis** [mortendahl/tf-encrypted#164]

**Comment by yanndupis** [comment]
@mortendahl @jvmancuso here is a quick status:

I have trained the different models with tensorflow then exported them as protobuf file and finally made private predictions with tfe on a sample size of 100 images. I have trained these models in plaintext completely separately from the tfe predictions (e.g. [network_a.py](https://github.com/mortendahl/tf-encrypted/blob/master/examples/securenn/network_a.py)) to check if we had an issue or not with our plaintext training (Tensorlfow). Here are the preliminary results:

**Network a**: Tensorflow and tfe predictions agree at 97%
**Network b**: Tensorflow and tfe predictions agree at 75%
**Network c**: Tensorflow and tfe predictions agree at 70% (can go down sometimes to 44% when re-trained)
**Network d"**: I think I have an issue with my implementation, the model was returning a matrix of shape [3,10].

Based on these observations, I think we have two issues:
- We have an issue in our plaintext training in the code located in the folder [/examples/securenn](https://github.com/mortendahl/tf-encrypted/tree/master/examples/securenn) or maybe when we transfer the weight to make private prediction. For example, when we execute the code network_a.py, the accuracy is around 10/20% however with the approach above I get 97%.
- We have an issue with the relu approximation. The accuracy decreases when the complexity of the model increases. Currently in tfe, we have approximated relu on the interval [-9,9]. The errors compound when we are outside this range or in the middle. 

Here are the distribution outputs for each layer before relu for the **network c**. These distributions are based on the model trained in plaintext:
<img width="718" alt="screen shot 2018-09-21 at 4 48 27 pm" src="https://user-images.githubusercontent.com/8001016/45910877-adfa4d00-bdc1-11e8-9d5d-e4e094254b59.png">
<img width="693" alt="screen shot 2018-09-21 at 4 49 00 pm" src="https://user-images.githubusercontent.com/8001016/45910881-b9e60f00-bdc1-11e8-9d86-f0b5c98f7720.png">

As we can see, some of the layer outputs before Relu are outside the expected range.

**Comment by yanndupis** [comment]
Thank you @jvmancuso ! I will have to clean up the code a bit before pushing it. I have added a branch called securenn_models for now.

**Comment by mortendahl** [comment]
This is amazing Yann, brilliant! Away this weekend but very much looking
forward to digging deeper on Monday 🙌

**Comment by yanndupis** [comment]
Thank you @mortendahl! Looking forward to Monday as well.

**Comment by yanndupis** [comment]
quick update:

I am trying to train the models with batchnorm before relu to see if it will improve the accuracy. I am experiencing some challenges to export the model from tf to a .pb file then import it into tf-encrypted. 

I trained couple weeks ago the network b with Pytorch. This model was transferred to tensorflow then exported to a .pb. It has batchnom before each relu. When comparing the output from tfe and tf, they agreed at 25%. vs 75% without batchnorm :(. I have found the problem. It's because relu is approximated on the interval [-9,9] so we loose precision in the middle. This has a bigger impact with batchnorm because the values are centered around 0. 
When I re-approximate relu on the interval [-3,3], for the network b with batchnorm, tfe and tf agree at 94%.

**Comment by yanndupis** [comment]
I think I was finally able to train the models with batchnorm in Tensorflow then export them in .pb file make predictions with tfe. Here are the preliminary results on the same images (sample size:100).

**Network a**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 100%
- **Relu Approximated [-9,9]**: Tensorflow and tfe predictions agree at 97%

**Network b**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 100%
- **Relu Approximated [-9,9]**: Tensorflow and tfe predictions agree at 90%

**Network c**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 100%
- **Relu Approximated [-9,9]**: Tensorflow and tfe predictions agree at 88%

So adding batchnorm before each relu seems to help, especially with relu is approximated on a smaller interval ( .e.g. [-3,3]). However, I am not sure that these results are always consistent depending how the model converged. I will have to re-train the model several time to validate. 

Thanks

**Comment by yanndupis** [comment]
I have compared the predictions with tf vs tfe on the full MNIST test dataset (10,000 images). With the Relu approximated on the intervall [-3,3] and batchnorm, here are the results:

**Network a**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 98.5%

**Network b**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 98.5%

**Network c**: 
- **Relu Approximated [-3,3]**: Tensorflow and tfe predictions agree at 99.5%

**Comment by ianlivingstone** [comment]
Beautiful!

**Comment by yanndupis** [comment]
Quick update on this. By merging the pr #234, it should solve most of our problems. In this example we should use Securenn protocol so we have the exact relu and won't have to add batchnorm before each relu. We also had some issues with plaintext training. All the drastic simplification Morgan introduced should solve most of our issues. Thank you.
